botcore/utils/function.py

```diff
@@ -0,0 +1,116 @@
+"""Utils for manipulating functions."""
+
+from __future__ import annotations
+
+import functools
+import types
+import typing
+from collections.abc import Sequence, Set
+from typing import Callable  # sphinx-autodoc-typehints breaks with collections.abc.Callable
+
+__all__ = ["command_wraps", "GlobalNameConflictError", "update_wrapper_globals"]
+
+
+if typing.TYPE_CHECKING:
+    import typing_extensions
+    _P = typing_extensions.ParamSpec("_P")
+    _R = typing.TypeVar("_R")
```

Comment:
> They don't need to be defined at runtime because they're only in annotations.

Tested it out. More specifically, it's fine because `from __future__ import annotations` is used.

Gonna leave this open for more perspectives on whether they should be public.